src/Collections/Arrays/BinarySearch.dfy

```diff
@@ -0,0 +1,41 @@
+// RUN: %dafny /compile:0 "%s"
+
+/*******************************************************************************
+*  Copyright by the contributors to the Dafny Project
+*  SPDX-License-Identifier: MIT 
+*******************************************************************************/
+
+include "../../StandardLibrary.dfy"
+
+module Search {
+  export
+    provides BinarySearch, Wrappers, StandardLibrary
+  
+  import StandardLibrary
+  import opened Wrappers = StandardLibrary.Wrappers
+
+  method BinarySearch<T>(a: array<T>, key: T, compare: (T, T) -> bool) returns (r:Option<nat>)
```

Comment:
Compare function usually return an int: negative for <, zero for == and positive for >. Or a datatype
If you a returning a bool, you should name this function "lessThan", "lessEqThan", "greaterThan" or "greaterEqThan", depending on what this means.

But my preference is really that you return a datatype [like this](https://github.com/dafny-lang/compiler-bootstrap/commit/d9a5734a47b9dd78b3af764ea913e7a19bc8c857#diff-2abd45eb58cd25c1bff3a753036307c5bc25d3309d470e9e45ea73d0308323acR588) and use the result in an

```
datatype ComparisonResult = LessThan | Equals | GreaterThan

var comp :=  compare(a[mid] , key);
match comp {
case LessThan =>
case Equals =>
case GreaterThan =>
}
```

Reply:
This again